collector/systemd_linux.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nosystemd
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type systemdCollector struct {
+	unitDesc          *prometheus.Desc
+	systemRunningDesc *prometheus.Desc
+}
+
+var unitStatesName = []string{"active", "activating", "deactivating", "inactive", "failed"}
+
+func init() {
+	Factories["systemd"] = NewSystemdCollector
+}
+
+// Takes a prometheus registry and returns a new Collector exposing
+// systemd statistics.
+func NewSystemdCollector() (Collector, error) {
+	const subsystem = "systemd"
+
+	unitDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, subsystem, "unit_state"),
+		"Systemd unit", []string{"name", "state"}, nil,
+	)
+	systemRunningDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, subsystem, "system_running"),
+		"Whether the system is operational (see `systemctl is-system-running')",
+		nil, nil,
+	)
+
+	return &systemdCollector{
+		unitDesc:          unitDesc,
+		systemRunningDesc: systemRunningDesc,
+	}, nil
+}
+
+func (c *systemdCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	units, err := c.listUnits()
+	if err != nil {
+		return fmt.Errorf("couldn't get units states: %s", err)
+	}
+	c.collectUnitStatusMetrics(ch, units)
+
+	systemState, err := c.getSystemState()
+	if err != nil {
+		return fmt.Errorf("couldn't get system state: %s", err)
+	}
+	c.collectSystemState(ch, systemState)
+
+	return nil
+}
+
+func (c *systemdCollector) collectUnitStatusMetrics(ch chan<- prometheus.Metric, units []dbus.UnitStatus) {
+	for _, unit := range units {
+		for _, stateName := range unitStatesName {
+			isActive := 0.0
+			if stateName == unit.ActiveState {
+				isActive = 1.0
+			}
+			ch <- prometheus.MustNewConstMetric(
+				c.unitDesc, prometheus.GaugeValue, isActive,
+				unit.Name, stateName)
+		}
+	}
+}
+
+func (c *systemdCollector) collectSystemState(ch chan<- prometheus.Metric, systemState string) {
+	isSystemRunning := 0.0
+	if systemState == "\"running\"" {
+		isSystemRunning = 1.0
+	}
+	ch <- prometheus.MustNewConstMetric(c.systemRunningDesc, prometheus.GaugeValue,
+		isSystemRunning)
```

Comment:
No need for this kind of line break here, as in Go there's usually no strict limit. It's more common to either put all args on a single line or put them each on their own line if the list gets very long. Single line should be fine here.